content/kotsadm/updating/patching-with-kustomize.md

```diff
@@ -0,0 +1,93 @@
+---
+date: 2019-10-09
+linktitle: "Patching with Kustomize"
+title: Patching with Kustomize
+weight: 10040
+draft: false
+---
+
+[Kustomize](https://kustomize.io) allows for last-mile patches (i.e. modifications) to the app that are otherwise not configurable via `Config` page.
+
+## Directory Structure
+
+Click on `View files` and lets take a look at the directory structure.
+![Kustomize Dir Structure](/images/kustomize-dir-structure.png)
+
+### Upstream
+
+The `upstream` dir mirrors exactly the content pushed to a release. This includes the template functions, preflight checks, support-bundle, config options, license etc.
+
+**Note:** No changes should be made in this dir as they are overwritten on each new release.
+
+### Base
+
+After KOTS processes and renders the `upstream` it puts those files in the `base` dir. Any non-deployable manifests such as template functions, preflight checks, config options, etc. are removed and only the "deployable" application (i.e. `kubectl apply`) will be placed here.
+
+**Note:** No changes should be made in this dir as they are overwritten on each new release.
+
+### Overlays
+
+The `overlays` dir references the `base` dir and this is where your local kustomize patches should be placed. Any changes made here will persist between releases unlike `upstream` and `base`.
+
+* * *
+
+## Patching the Overlays
+
+Since `upstream` and `base` are ephemeral, lets make a kustomize patch in `overlays/downstreams/this-cluster/` so it persists between releases.
+
+Download the KOTS app locally.
+
+```shell
+kubectl kots download --namespace <namespace> --slug <app_slug> --dest ~/my-kots-app
+  • Connecting to cluster ✓
+
+```
+
+Lets patch something simple like the `replicas` count. Create a file in `~/my-kots-app/overlays/downstreams/this-cluster/patch-deployment.yaml` with the following.
+
+```shell
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-nginx
+spec:
+  replicas: 2
+```
+
+Don't forget to add this file under `patches` in `~/my-kots-app/overlays/downstreams/this-cluster/kustomization.yaml`
+
+```shell
+apiVersion: kustomize.config.k8s.io/v1beta1
+bases:
+- ../../midstream
+kind: Kustomization
+patches:
```

Comment:
Did you try this with `patchesStrategicMerge` and it didn't work?

Reply:
What do you think about showing the change here with a `diff` type patch, e.g. with `diff -Naur`

```diff
--- k1.yaml     2020-04-14 10:29:00.000000000 -0500
+++ k2.yaml     2020-04-14 10:29:21.000000000 -0500
@@ -2,3 +2,5 @@
 bases:
 - ../../midstream
 kind: Kustomization
+patches:
+- ./patch-deployment.yaml
```

(click "edit this comment to see the markdown source I think)